File: packages/core/src/core/toolPackMapping.ts
/**
 * Tool name → OAP policy pack ID. Single source: tool-pack-mapping.json.
 * Used by evaluator and exported for adapters (LangChain, CrewAI) so they pass the correct capability.
 */

import path from 'node:path';
import { readFileSync } from 'node:fs';

interface MappingRule {
  prefixes?: string[];
  substrings?: string[];
  pack: string;
}

declare const __dirname: string;

const data: { default: string; rules: MappingRule[] } = JSON.parse(
  readFileSync(path.join(__dirname, 'tool-pack-mapping.json'), 'utf8')
);

/**
 * Map a tool name to the OAP policy pack ID. Single source of truth from tool-pack-mapping.json.
 */
export function toolToPackId(toolName: string): string {
  const t = (toolName ?? '').trim().toLowerCase();
  for (const rule of data.rules) {
    if (rule.prefixes?.some((pre) => t.startsWith(pre))) return rule.pack;
    if (rule.substrings?.some((sub) => t.includes(sub))) return rule.pack;
  }
  return data.default;
}

File: python/aport_guardrails/core/tool_pack_mapping.py
"""
Tool name → OAP policy pack ID. Single source: tool-pack-mapping.json.
Canonical file: packages/core/src/core/tool-pack-mapping.json (sync this copy when updating).
"""

import json
from pathlib import Path

_MAPPING_DIR = Path(__file__).resolve().parent
_CACHED: dict | None = None


def _load_mapping() -> dict:
    global _CACHED
    if _CACHED is not None:
        return _CACHED
    p = _MAPPING_DIR / "tool-pack-mapping.json"
    _CACHED = json.loads(p.read_text())
    return _CACHED


def tool_to_pack_id(tool_name: str) -> str:
    """Map a tool name to the OAP policy pack ID. Single source: tool-pack-mapping.json."""
    t = (tool_name or "").strip().lower()
    data = _load_mapping()
    default = data.get("default", "system.command.execute.v1")
    for rule in data.get("rules", []):
        if rule.get("prefixes") and any(t.startswith(pre) for pre in rule["prefixes"]):
            return rule["pack"]
        if rule.get("substrings") and any(sub in t for sub in rule["substrings"]):
            return rule["pack"]
    return default
